src/type-data-frame.c

```diff
@@ -353,10 +380,65 @@ SEXP df_ptype2(SEXP x, SEXP y, struct vctrs_arg* x_arg, struct vctrs_arg* y_arg)
 
   init_data_frame(out, 0);
 
-  UNPROTECT(6);
+  UNPROTECT(4);
+  return out;
+}
+
+SEXP df_ptype2_sequential(SEXP x, SEXP y,
+                          SEXP names, R_len_t len,
+                          struct vctrs_arg* x_arg, struct vctrs_arg* y_arg) {
+  SEXP out = PROTECT(Rf_allocVector(VECSXP, len));
+  Rf_setAttrib(out, R_NamesSymbol, names);
+
+  for (R_len_t i = 0; i < len; ++i) {
+    const char* name = r_chr_get_c_string(names, i);
+
+    struct arg_data_index x_arg_data = new_index_arg_data(name, x_arg);
+    struct arg_data_index y_arg_data = new_index_arg_data(name, y_arg);
+    struct vctrs_arg named_x_arg = new_index_arg(x_arg, &x_arg_data);
+    struct vctrs_arg named_y_arg = new_index_arg(y_arg, &y_arg_data);
+
+    int _left;
+
+    SEXP type = vec_ptype2(VECTOR_ELT(x, i),
+                           VECTOR_ELT(y, i),
+                           &named_x_arg,
+                           &named_y_arg,
+                           &_left);
+
+    SET_VECTOR_ELT(out, i, type);
+  }
+
+  init_data_frame(out, 0);
+
+  UNPROTECT(1);
   return out;
 }
 
+static bool identical_names(SEXP x_names, SEXP y_names, R_len_t x_len, R_len_t y_len) {
```

Comment:
Can you try with `equal_object()` and see if the performance is similar?

If this helper is needed it should go in equal.c and use the templated loop `EQUAL_ALL()`. That should also improve the performance for the case of equal names with different encoding.

Reply:
Ah `equal_object()` seems fine. I also tweaked it to use the optimized `*_na_equal()` functions since we hardcode `na_equal = true` here